### ISSUE TITLE
Do we need administrative controllers?

### ISSUE THREAD

**Issue by phproberto** [icarus/logical-j#1]
Current ACL system already checks for user permissions. I don't see why we have to restrict specific actions to backend. I'm talking about:

https://github.com/icarus/logical-j/blob/MVSC/libraries/cms/controller/add.php#L12

I think we need a single way of doing things in both front and backend. I may add an item in frontend, I may delete an item in frontend.

Components already can have a global permission check for manage in the entry point. Talking about it:

https://github.com/icarus/logical-j/blob/MVSC/libraries/cms/controller/administrate.php#L23-L26

Move that to the component entry point as a fast check for global backend access and keep controllers logic relay only in the ACL.

**Comment by Mathewlenning** [comment]
Actually none of these controllers are restricted to back-end. The only difference between administrative controllers and other controllers is that the user has to have core.manage access to even use them. 

I use them for both front and back editing using the same model, same view, same layouts everything. MVSC controllers are fully auto loaded so they don't know anything about the file structure.

**Comment by phproberto** [comment]
Yes but let's create an example. I have an agent that can create/edit clients in frontend. Or I may let my users to edit their own profile in frontend. I don't want them to have access to my backend or even core.manage rights.

**Comment by Mathewlenning** [comment]
So you give them access to core.edit and core.manage for the profile component, not for the the admin component.

**Comment by Mathewlenning** [comment]
Or you create a new controller that addresses that specific need. 

Then you can reuse it in any extension you have from that day forward.

**Comment by Mathewlenning** [comment]
The ACL is by component, so setting access for editing/managing in say com_content won't allow them to login to the administration area.

**Comment by Mathewlenning** [comment]
The reason for the administrative controller is that component ACL is for better or worst up to the extension developer. If they choose not to implement it, we don't want CRUD to be open to the world.

**Comment by phproberto** [comment]
There is already an ACL system that works great. In fact in my extensions I'm using same controllers for everything. Having that core.manage thing there implies that I have to start overriding things. And it's not really needed. 

You are thinking in different component for profile or the rest but it can be the same component. So I just need to add an action to the ACL system. 

For me it's quite simple core.add is to allow to add core.edit is to allow to edit, etc. Making core.manage mandatory makes no sense at all here and forces extension developers to start creating their own shit. And I won't be able to call `parent::_construtct()` if that breaks my ACL so I'll need to start creating my own layer over this and we will end with 5 frameworks over MVSC :dancer: .

**Comment by mbabker** [comment]
You're adding an inherent ACL dependency that all users have to have `core.manage

**Comment by mbabker** [comment]
In truth, general view type access checks should be done in a component dispatcher (the component entry point files) and more granular checks done at the level needed.  `core.manage` is essentially a basic view permission.

**Comment by Mathewlenning** [comment]
Well just to be clear, I'm not trying to copy the CMS here and yes I am creating a requirement that any administrative task requires at minimum core.manage.

But core.manage isn't used for "View Access" in MVSC that is handled by the views canView method. In MVSC GET is unprotected. You can get whatever you want, so long as the view says it is o.k.
 
I should have mentioned that in MVSC the ACL hierarchy is a little different.
In Legacy the hierarchy is something like Global -> Component -> Category -> Record
In MVSC the hieracrchy is Global -> Component -> Resource -> Record

Long and short you can set ACL per model.

**Comment by Mathewlenning** [comment]
Before you start thinking I hacked the core ACL, I didn't touch the code or the DB schema. I just added a controller that will save an asset with the format component.resource

Then in my table I check for component.resource before checking for component when looking for the parent.

**Comment by mbabker** [comment]
:+1:

**Comment by Mathewlenning** [comment]
Another point that should probably be made is that ACL checks are not done in the model. The assumption is if you made it through the controllers, to the model method you're good to go.

Why? Because if you want to use your model to do something besides CRUD then having ACL checks really becomes a problem. And if someone has hacked your site and has placed a program on your server that can access your models directly, you've got bigger problems to worry about.

Especially since they could just load the JConfig and get the DB credentials and access the data directly.

**Comment by Mathewlenning** [comment]
@phproberto one option that probably none of the other proposals so far hasn't really offered is the ability to override the dispatcher controller, so that it uses the same type of logic used for routing in the CMS now. 

What I mean is that you could use your current multi-task controller in MVSC without a bunch of overriding or 5 frameworks. 

The models and views don't know about the controllers, so they don't care who calls them. You would probably have to pull the get view method into your controller, but that is only one method.

**Comment by Mathewlenning** [comment]
Also since each controller is given full authority over the meaning of all the configuration variables, whatever meaning you decided to give to them is up to you.

**Comment by phproberto** [comment]
> Well just to be clear, I'm not trying to copy the CMS here and yes I am creating a requirement that any administrative task requires at minimum core.manage.

I thought this was to build a common system for v4 and share thoughts about. I took the last sentence as "this is my way of doing things". If so I think I can save my comments and waste my time in other things.

> I should have mentioned that in MVSC the ACL hierarchy is a little different.
In Legacy the hierarchy is something like Global -> Component -> Category -> Record
In MVSC the hieracrchy is Global -> Component -> Resource -> Record

Standard ACL si simple. Check standard permission assigned to that action and done. Then if you have categories you need to override the canEdit canCreate and so.

I'm not sure that changing how ACL works will make this a too big jump and we are going to lose more extension developers.

> Another point that should probably be made is that ACL checks are not done in the model. The assumption is if you made it through the controllers, to the model method you're good to go.

I agree. ACL checks have to be done in the controllers. The current code in models is because someone didn't find a better way to do it properly.

It would be good to have a sample component using this new system. For now I see that I'll need to start overriding too many things.

**Comment by Mathewlenning** [comment]
Actually the logical-j branch is about building my version of what "v4" could be.

Icarus is about finding the best option  out of the various possible versions. 

I do appreciate the input. It is hard to address concerns when I don't know what they are. 

But if you feel like it's a waste of time, then maybe opening you're own Icarus repo and building your version of the future would be more productive.

Sincerely,
Mathew Lenning

Babel-university.com

P.S. This message was sent via iPhone, so please forgive any errors
